src/Starlight/Draw/Starfield.hs

```diff
@@ -84,29 +87,37 @@ range = Interval 0 (Identity (length vertices))
 -- based on Star Nest by Pablo Roman Andrioli: https://www.shadertoy.com/view/XlfGRj
 
 shader :: Shader U V O
-shader = program $ \ U{ resolution, origin, zoom }
+shader = program $ \ U{ resolution, focusQ, focusR, zoom }
   ->  vertex (\ V{ pos } None -> main $
     gl_Position .= ext4 (ext3 pos 0) 1)
 
   >>> fragment (\ None O{ fragColour } -> main $ do
-    uv <- let' "uv" $ (gl_FragCoord ^. _xy / resolution ^. _xy - 0.5) * vec2 1 (resolution ^. _y / resolution ^. _x)
+    resolution <- let' @(V2 Float) "resolution" (D.coerce resolution)
+    uv <- let' "uv" $ (gl_FragCoord^._xy / resolution^._xy - 0.5) * vec2 [1, resolution^._y / resolution^._x]
     dir <- var "dir" $ ext3 (uv D.^* zoom) 1 D.^* 0.5
-    origin <- var "origin" $ ext3 (D.coerce origin / (resolution D.^* 0.1)) 1
-    a1 <- let' "a1" $ 0.2 + norm (get origin) / resolution ^. _x * 2
-    a2 <- let' "a2" $ 0.2 + norm (get origin) / resolution ^. _y * 2
-    rot1 <- let' "rot1" $ mat2 (vec2 (cos a1) (sin a1)) (vec2 (-(sin a1)) (cos a1))
-    rot2 <- let' "rot2" $ mat2 (vec2 (cos a2) (sin a2)) (vec2 (-(sin a2)) (cos a2))
-    dir ^^. _xz *!= rot1
-    dir ^^. _xy *!= rot2
-    origin ^^. _xz *!= rot1
-    origin ^^. _xy *!= rot2
+    focusR <- var @(V3 Float) "focusR" $ ext3 focusR 1
+    focusQ <- var @(V3 Float) "focusQ" $ ext3 focusQ 1
+    let wrap mn mx x = ((x + mx) `mod'` (mx - mn)) + mn
+    nf <- let' "nf" (norm (get focusQ + get focusR))
+    a1 <- let' "a1" $ wrap (-pi) pi (0.3 + 0.1/nf)
+    a2 <- let' "a2" $ wrap (-pi) pi (0.2 + 0.1/nf)
+    rot1 <- let' "rot1" $ mat2 (vec2 [cos a1, sin a1]) (vec2 [-sin a1, cos a1])
+    rot2 <- let' "rot2" $ mat2 (vec2 [cos a2, sin a2]) (vec2 [-sin a2, cos a2])
+    dir^^._xz *!= rot1
+    dir^^._xy *!= rot2
+    focusR^^._xz *!= rot1
+    focusR^^._xy *!= rot2
+    focusQ^^._xz *!= rot1
+    focusQ^^._xy *!= rot2
+    focusR *= vec3 [10]
+    focusQ *= vec3 [10]
     s <- var "s" 0.1
     fade <- var "fade" 0.5
-    v <- var "v" $ vec3 0 0 0
+    v <- var "v" $ vec3 [0]
     r <- var @Int "r" 0
     while (get r `lt` volsteps) $ do
-      p <- var "p" $ get origin + get dir D.^* get s
-      p .= abs (tile - mod' (get p) (tile D.^* 2))
+      p <- var "p" $ get focusR + get focusQ + get dir D.^* get s
+      p .= abs (vec3 [tile] - mod' (get p) (vec3 [tile * 2]))
```

Comment:
…so that they can ultimately be recombined.

This results in the starfield being much less jittery/janky at long distances from the sun.